### ISSUE TITLE
Breaking changes in `object_changes` deserialization for `ActiveRecord::Enum`?

### ISSUE THREAD

**Issue by hsluo** [airblade/paper_trail#855]
We recently upgraded paper_trail from v4.0.2 to the latest version for Rails 5 migration, and find that there's something unexpected when reading `changeset` from `version`.

As of v4.0.2, changes of enums are stored as strings. For example, with the `PostWithStatus` model

```ruby
post = PostWithStatus.create(status: :draft)
post.published! # version.object_changes should contain "---\nstatus:\n- draft\n- published"
post.versions.last.changeset #=> {"status"=>["draft","published"]}
```

Later a new serialization method was introduced which serializes enum to integer, and the changeset deserialization for older versions is broken. 

```ruby
post.versions.last.changeset #=> {"status"=>["draft","draft"]}
```

I think it's because the values are assumed to be integers, and it will be parsed to integers and used to reverse lookup in the enum mapping. `'string'.to_i` returns 0, so values in the changeset will all be the 0th value of the enum.

I've tested with Rails 4 and 5 and think it's an issue of paper_trail. Please let me know if there's something I'm doing wrong or missing.

**Comment by jaredbeck** [comment]
This may be a duplicate of https://github.com/airblade/paper_trail/issues/798, which has been fixed, but not released yet. Please try the `master` branch and let me know if it fixes your issue, thanks.

**Comment by hsluo** [comment]
No, it's not fixed.

**Comment by jaredbeck** [comment]
OK, thanks for testing. Please provide a failing test using our [bug report template][1]. Thanks.

[1]: https://github.com/airblade/paper_trail/blob/master/doc/bug_report_template.rb

**Comment by hsluo** [comment]
Here's the bug report.
https://gist.github.com/hsluo/312d7c2bd9e062f67ea3e3a3d5a7334f

**Comment by jaredbeck** [comment]
Thanks for the bug report, very helpful. I think we can incorporate your failing test into `spec/models/post_with_status_spec.rb`, and I'm guessing the fix will involve `attribute_serializers/cast_attribute_serializer.rb`.

The first idea that comes to mind is to do something ugly like this:

```diff
         def deserialize(attr, val)
-          @klass.type_for_attribute(attr).deserialize(val)
+          if defined_enums[attr] && val.is_a?(::String)
+            # Because PT 4 used to save the string version of enums to `object_changes`
+            val
+          else
+            @klass.type_for_attribute(attr).deserialize(val)
+          end
+        end
```

It sure feels like a hack, though. I hope someone else has a better idea :)

**Comment by jaredbeck** [comment]
Fixed in #857, released in 5.2.1. Thanks @hsluo

**Comment by hsluo** [comment]
Thanks!
